tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -0,0 +1,329 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for lowering LHLO dialect to Affine dialect.
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/mhlo/IR/lhlo_ops.h"
+#include "mlir-hlo/Dialect/mhlo/transforms/fusion_utils.h"
+#include "mlir-hlo/Dialect/mhlo/transforms/lhlo_elemental_utils.h"
+#include "mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h"
+#include "mlir-hlo/utils/codegen_utils.h"
+#include "mlir-hlo/utils/placement_utils.h"
+#include "mlir/Dialect/GPU/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+
+using mlir::codegen_utils::calcMultiDimIndex;
+using mlir::memref::DimOp;
+using mlir::memref::LoadOp;
+using mlir::memref::StoreOp;
+
+namespace mlir {
+namespace lmhlo {
+
+#define GEN_PASS_CLASSES
+#include "mlir-hlo/Dialect/mhlo/transforms/lmhlo_passes.h.inc"
+
+namespace {
+
+template <typename LHLO_OpTy>
+bool elemwiseLowerHelper(
+    OpBuilder& b, Location loc, Operation* op, Value output_linear_index,
+    const ShapeConstraintAnalysis* shape_constraint_analysis) {
+  if (!isa<LHLO_OpTy>(op)) {
+    return false;
+  }
+  Value result_memref = cast<LmhloOp>(op).getResultBuffer();
+  Value memref = result_memref;
+  if (shape_constraint_analysis) {
+    Value leader_memref =
+        shape_constraint_analysis->GetLeaderValueWithSameShape(result_memref);
+    if (leader_memref != nullptr) {
+      memref = leader_memref;
+    }
+  }
+  // TODO(disc): Replace with memref.Delinearize
+  auto multidim_index = calcMultiDimIndex(b, loc, output_linear_index, memref);
+  SmallVector<Value, 4> operand_values;
+  int num_operands = op->getNumOperands();
+  for (int i = 0; i < num_operands - 1; ++i) {
+    auto operand_memref = *(op->getOperands().begin() + i);
```

Comment:
```suggestion
  for (Value operand_memref : op->getOperands().drop_back());
```

You're not using `i` in the loop.

Reply:
done